File: core/src/com/qtech/bubbleblaster/registry/LocaleManager.java
package com.qtech.bubbleblaster.registry;

import com.qtech.bubbleblaster.common.maps.SequencedHashMap;
import com.qtech.bubbleblaster.common.text.translation.LanguageMap;

import java.util.*;

public final class LocaleManager extends AbstractRegistry<Locale, LanguageMap> {
    //    protected final static HashBasedTable<Object, Object, Object> registries = HashBasedTable.create();
    protected final static SequencedHashMap<Locale, LanguageMap> registry = new SequencedHashMap<>();

    public static LocaleManager getManager() {
        return new LocaleManager();
    }

    protected LocaleManager() throws IllegalStateException {

    }

    /**
     * Returns the registered instance from the given {@link Locale}
     *
     * @param obj the namespaced key.
     * @return an registered instance of the type {@link LanguageMap}.
     * @throws ClassCastException if the type is invalid.
     */
    public LanguageMap get(Locale obj) {
        return registry.get(obj);
    }

    /**
     * Register an object.
     *
     * @param key the register item key.
     * @param val the register item value.
     */
    @Override
    @Deprecated
    public void register(Locale key, LanguageMap val) {
        registry.put(key, val);
    }

    public void register(String locale) {
        Locale locale1 = Locale.forLanguageTag(locale);
        LanguageMap languageMap = new LanguageMap(locale1);
        register(locale1, languageMap);
    }

    @Override
    public Collection<LanguageMap> values() {
        return new ArrayList<>(registry.values());
    }

    @Override
    public Set<Locale> keys() {
        return registry.keySet();
    }

    @Override
    public Set<Map.Entry<Locale, LanguageMap>> entries() throws IllegalAccessException {
        // I do this because IDE won's accept dynamic values ans keys.
        ArrayList<LanguageMap> values = new ArrayList<>(values());
        ArrayList<Locale> keys = new ArrayList<>(keys());

        if (keys.size() != values.size()) throw new IllegalAccessException("Keys and values have different lengths.");

        Set<Map.Entry<Locale, LanguageMap>> entrySet = new HashSet<>();

        for (int i = 0; i < keys.size(); i++) {
            entrySet.add(new AbstractMap.SimpleEntry<>(keys.get(i), values.get(i)));
        }

        return entrySet;
    }
}

File: src/main/kotlin/qtech/bubbles/registry/LocaleManager.kt
@file:Suppress("DEPRECATION")

package qtech.bubbles.registry

import qtech.bubbles.common.maps.SequencedHashMap
import qtech.bubbles.common.text.translation.LanguageMap
import java.util.*
import java.util.AbstractMap.SimpleEntry

open class LocaleManager protected constructor() : AbstractRegistry<Locale, LanguageMap?>() {
    /**
     * Returns the registered instance from the given [Locale]
     *
     * @param obj the namespaced key.
     * @return an registered instance of the type [LanguageMap].
     * @throws ClassCastException if the type is invalid.
     */
    override fun get(obj: Locale): LanguageMap? {
        return Companion.registry[obj]
    }

    /**
     * Register an object.
     *
     * @param key the register item key.
     * @param val the register item value.
     */
    @Deprecated("", ReplaceWith("register(locale)"))
    override fun register(key: Locale, `val`: LanguageMap?) {
        Companion.registry[key] = `val`
    }

    fun register(locale: String?) {
        val locale1 = Locale.forLanguageTag(locale)
        val languageMap = LanguageMap(locale1)
        register(locale1, languageMap)
    }

    override fun values(): ArrayList<LanguageMap?> {
        return ArrayList(Companion.registry.values)
    }

    override fun keys(): Set<Locale> {
        return Companion.registry.keys
    }

    @Throws(IllegalAccessException::class)
    override fun entries(): Set<Map.Entry<Locale, LanguageMap?>> {
        // I do this because IDE won's accept dynamic values ans keys.
        val values = ArrayList(values())
        val keys = ArrayList(keys())
        if (keys.size != values.size) throw IllegalAccessException("Keys and values have different lengths.")
        val entrySet: MutableSet<Map.Entry<Locale, LanguageMap?>> = HashSet()
        for (i in keys.indices) {
            entrySet.add(SimpleEntry(keys[i], values[i]))
        }
        return entrySet
    }

    companion object {
        //    protected final static HashBasedTable<Object, Object, Object> registries = HashBasedTable.create();
        protected val registry = SequencedHashMap<Locale, LanguageMap>()
        @JvmStatic
        val manager: LocaleManager
            get() = LocaleManager()
    }
}